test/support/sycl_alloc_utils.h

```diff
@@ -126,40 +150,109 @@ class usm_data_transfer
         return __ptr;
     }
 
-    template<typename _Iterator>
+    template <typename _Iterator>
+    void update_data(_Iterator __it)
+    {
+        update_data(__it, 0, __count);
+    }
+
+    template<typename _Iterator, typename _Size>
+    void update_data(_Iterator __it, __difference_type __offset, _Size __objects_count)
+    {
+        assert(0 <= __offset);
+        assert(0 <= __objects_count);
+        assert(__offset + __objects_count <= __count);
+
+        if (__count > 0 && __objects_count > 0)
+        {
+            assert(__ptr != nullptr);
+
+            //TODO: support copying data provided by non-contiguous iterator
+            auto __src = std::addressof(*__it);
+            assert(std::addressof(*(__it + __objects_count)) - __src == __objects_count);
+
+            copy_data_impl(__src, __ptr + __offset, __objects_count);
+        }
+    }
+
+    template <typename _Iterator>
     void retrieve_data(_Iterator __it)
     {
-        if (__count > 0)
+        return retrieve_data(__it, 0, __count);
+    }
+
+    template<typename _Iterator>
+    void retrieve_data(_Iterator __it, __difference_type __offset, __difference_type __objects_count)
+    {
+        assert(0 <= __offset);
+        assert(0 <= __objects_count);
+        assert(__offset + __objects_count <= __count);
+
+        if (__count > 0 && __objects_count > 0)
         {
             assert(__ptr != nullptr);
 
             //TODO: support copying data provided by non-contiguous iterator
             auto __dst = std::addressof(*__it);
-            assert(std::addressof(*(__it + __count)) - __dst == __count);
+            assert(std::addressof(*(__it + __objects_count)) - __dst == __objects_count);
+
+            copy_data_impl(__ptr + __offset, __dst, __objects_count);
+        }
+    }
 
+private:
+
+    void copy_data_impl(_ValueType* __src, _ValueType* __ptr, __difference_type __count)
+    {
 #if __LIBSYCL_VERSION >= 50300
-            __queue.copy(__ptr, __dst, __count);
+        __queue.copy(__src, __ptr, __count);
 #else
-            auto __p = __ptr;
-            auto __c = __count;
-            __queue.submit([__dst, __c, __p](sycl::handler& __cgh){
-                __cgh.parallel_for(sycl::range<1>(__c), [__dst, __c, __p](sycl::item<1>__item){
-                    ::std::size_t __id = __item.get_linear_id();
-                    *(__dst + __id) = *(__p + __id);
+        auto __p = __ptr;
+        auto __c = __count;
+        __queue.submit([__src, __c, __p](sycl::handler& __cgh) {
+            __cgh.parallel_for(sycl::range<1>(__c), [__src, __c, __p](sycl::item<1>__item) {
+                ::std::size_t __id = __item.get_linear_id();
+                *(__p + __id) = *(__src + __id);
```

Comment:
@dmitriy-sobolev please fix this comment and merge into main branch.

Reply:
I found that this code is invalid because of accessing of host memory from within a kernel. The tests pass on CPU, but running them on GPU leads to CL_INVALID_ARG_VALUE. I am going to turn off testing with USM with old compilers, because I see no other way to copy data to host only memory.

Reply:
@dmitriy-sobolev Please ping me when you will finish this changes.